#Valeria Cortez Carreón A00571474
#A2_DataManagementIris
#Last updated: 14/09/2021

# Importing libraries


In [102]:
# Import the packages that we will be using
import pandas as pd                 # For data handling

# Importing data

In [103]:
# url string that hosts our .csv file
url = "datasets/iris/iris.csv"

# Read the .csv file and store it as a pandas Data Frame
df = pd.read_csv(url, names=['Sepal_length', 'Sepal_width', 'Petal_length', 'Petal_width', 'Species', 'Species_NumValue'])


If we want to print the information about the output object type we would simply type the following:

In [104]:
type(df)

pandas.core.frame.DataFrame

# Exploring the content of the data set


In [105]:
df.shape

(150, 6)

If we want to show the entire data frame we would simply write the following:

In [106]:
df

Sepal_length  Sepal_width  Petal_length  Petal_width         Species  \
0             5.1          3.5           1.4          0.2     Iris-setosa   
1             4.9          3.0           1.4          0.2     Iris-setosa   
2             4.7          3.2           1.3          0.2     Iris-setosa   
3             4.6          3.1           1.5          0.2     Iris-setosa   
4             5.0          3.6           1.4          0.2     Iris-setosa   
..            ...          ...           ...          ...             ...   
145           6.7          3.0           5.2          2.3  Iris-virginica   
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   

     Species_NumValue  
0                   1  
1                   1  
2                   1  
3                   1  
4                   1  
..                ...  
145                 3  
146                 3  
147                 3  
148                 3  
149                 3  

[150 rows x 6 columns]

`head()` function to show the first 5 rows of our data frame

In [107]:
df.head()
#df.head(10)

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

`tail()` function to show the last 5 rows of our data frame

In [108]:
df.tail()
#df.tail(3)

Sepal_length  Sepal_width  Petal_length  Petal_width         Species  \
145           6.7          3.0           5.2          2.3  Iris-virginica   
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   

     Species_NumValue  
145                 3  
146                 3  
147                 3  
148                 3  
149                 3

The columns in a Pandas data frame have names, to see the names, use the `columns` method:

To gather more information regarding the data, we can view the column names with the following function:

In [109]:
df.columns

Index(['Sepal_length', 'Sepal_width', 'Petal_length', 'Petal_width', 'Species',
       'Species_NumValue'],
      dtype='object')

Be aware that every variable in a Pandas data frame has a data type.  There are many different data types, but most commonly you will encounter floating point values (real numbers), integers, strings (text), and date/time values.  When Pandas reads a text/csv file, it guesses the data types based on what it sees in the first few rows of the data file.  Usually it selects an appropriate type, but occasionally it does not.  To confirm that the data types are consistent with what the variables represent, inspect the `dtypes` attribute of the data frame.

In [110]:
df.dtypes

Sepal_length        float64
Sepal_width         float64
Petal_length        float64
Petal_width         float64
Species              object
Species_NumValue      int64
dtype: object

Summary statistics, which include things like the mean, min, and max of the data, can be useful to get a feel for how large some of the variables are and what variables may be the most important. 

In [111]:
# Summary statistics for the quantitative variables
df.describe()


Sepal_length  Sepal_width  Petal_length  Petal_width  Species_NumValue
count    150.000000   150.000000    150.000000   150.000000        150.000000
mean       5.843333     3.057333      3.758000     1.199333          2.000000
std        0.828066     0.435866      1.765298     0.762238          0.819232
min        4.300000     2.000000      1.000000     0.100000          1.000000
25%        5.100000     2.800000      1.600000     0.300000          1.000000
50%        5.800000     3.000000      4.350000     1.300000          2.000000
75%        6.400000     3.300000      5.100000     1.800000          3.000000
max        7.900000     4.400000      6.900000     2.500000          3.000000

In [112]:
# Drop observations with NaN values
#Not NaN values found in data set
df.Sepal_length.dropna().describe()
#df.Wingspan.dropna().describe()

count    150.000000
mean       5.843333
std        0.828066
min        4.300000
25%        5.100000
50%        5.800000
75%        6.400000
max        7.900000
Name: Sepal_length, dtype: float64

It is also possible to get statistics on the entire data frame or a column as follows

- `df.mean()` Returns the mean of all columns
- `df.corr()` Returns the correlation between columns in a data frame
- `df.count()` Returns the number of non-null values in each data frame column
- `df.max()` Returns the highest value in each column
- `df.min()` Returns the lowest value in each column
- `df.median()` Returns the median of each column
- `df.std()` Returns the standard deviation of each column

In [113]:
df.mean()

Sepal_length        5.843333
Sepal_width         3.057333
Petal_length        3.758000
Petal_width         1.199333
Species_NumValue    2.000000
dtype: float64

In [141]:
df.median()

Sepal_length        5.8
Sepal_width         3.0
Petal_length        4.4
Petal_width         1.3
Species_NumValue    2.0
dtype: float64

# Write a data frame to a File

To save a file with your data simply use the `to_csv` attribute

In [114]:
df.to_csv('myDataFrameIris.csv')
#df.to_csv('myDataFrame.csv', sep='\t')

# Rename columns

To change the name of a colum use the `rename` attribute

In [115]:
df = df.rename(columns={"Species": "Iris Species"})

df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width Iris Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

In [116]:
df = df.rename(columns={"Iris Species": "Species"})

df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

# Selection of colums

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent observations or samples and the columns represent variables.  One common manipulation of a data frame is to extract the data for one case or for one variable.  There are several ways to do this, as shown below.

To extract all the values for one column (variable), use one of the following alternatives.

In [117]:
a = df.Petal_length
b = df["Petal_length"]
c = df.loc[:, "Petal_length"]
d = df.iloc[:, 2]

print(d)

0      1.4
1      1.4
2      1.3
3      1.5
4      1.4
      ... 
145    5.2
146    5.0
147    5.2
148    5.4
149    5.1
Name: Petal_length, Length: 150, dtype: float64


# Slicing a data set

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent cases and the columns represent variables.  One common manipulation of a data frame is to extract the data for one observation or for one variable.  There are several ways to do this, as shown below.

Lets say we would like to splice our data frame and select only specific portions of our data.  There are three different ways of doing so.

1. .loc()
2. .iloc()
3. .ix()

We will cover the .loc() and .iloc() splicing functions.


The attibute **.loc()** uses labels/column names, in specific, it takes two single/list/range operator separated by ',', the first one indicates the rows and the second one indicates columns.

In [118]:
# Select all rows for multiple columns, ["Gender", "GenderGroup"]
df.loc[:,["Species", "Species_NumValue"]]

# Select multiple columns, ["Gender", "GenderGroup"]me 
#keep = ["Species", "Species_NumValue"]
#df_sepal = df[keep]

# Select few rows for multiple columns, ["CWDistance", "Height", "Wingspan"]
df.loc[30:60, ["Species", "Species_NumValue"]]

# Select range of rows for all columns
#df.loc[10:15,:]



Species  Species_NumValue
30      Iris-setosa                 1
31      Iris-setosa                 1
32      Iris-setosa                 1
33      Iris-setosa                 1
34      Iris-setosa                 1
35      Iris-setosa                 1
36      Iris-setosa                 1
37      Iris-setosa                 1
38      Iris-setosa                 1
39      Iris-setosa                 1
40      Iris-setosa                 1
41      Iris-setosa                 1
42      Iris-setosa                 1
43      Iris-setosa                 1
44      Iris-setosa                 1
45      Iris-setosa                 1
46      Iris-setosa                 1
47      Iris-setosa                 1
48      Iris-setosa                 1
49      Iris-setosa                 1
50  Iris-versicolor                 2
51  Iris-versicolor                 2
52  Iris-versicolor                 2
53  Iris-versicolor                 2
54  Iris-versicolor                 2
55  Iris-versicolor                 2
56  Iris-versicolor                 2
57  Iris-versicolor                 2
58  Iris-versicolor                 2
59  Iris-versicolor                 2
60  Iris-versicolor                 2

The attribute **iloc()** is an integer based slicing.

In [119]:
# .
df.iloc[:, :5]

# .
df.iloc[:5, :]

# .
df.iloc[:, 3:7]

# .
df.iloc[4:8, 2:4]

# This is incorrect:
#df.iloc[1:5, ["Gender", "GenderGroup"]]

Petal_length  Petal_width
4           1.4          0.2
5           1.7          0.4
6           1.4          0.3
7           1.5          0.2

# Get unique existing values

List unique values in the one of the columns


In [120]:
# List unique values in the df['Species'] column
df.Species.unique()

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

# Filter, Sort and Groupby



With **Filter** you can use different conditions to filter columns. For example, df[df[year] > 1984] would give you only the column year is greater than 1984. You can use & (and) or | (or) to add different conditions to your filtering. This is also called boolean filtering.

In [121]:
#Petals longer than the average
df[df["Petal_length"] >= 3.7]

Sepal_length  Sepal_width  Petal_length  Petal_width          Species  \
50            7.0          3.2           4.7          1.4  Iris-versicolor   
51            6.4          3.2           4.5          1.5  Iris-versicolor   
52            6.9          3.1           4.9          1.5  Iris-versicolor   
53            5.5          2.3           4.0          1.3  Iris-versicolor   
54            6.5          2.8           4.6          1.5  Iris-versicolor   
..            ...          ...           ...          ...              ...   
145           6.7          3.0           5.2          2.3   Iris-virginica   
146           6.3          2.5           5.0          1.9   Iris-virginica   
147           6.5          3.0           5.2          2.0   Iris-virginica   
148           6.2          3.4           5.4          2.3   Iris-virginica   
149           5.9          3.0           5.1          1.8   Iris-virginica   

     Species_NumValue  
50                  2  
51                  2  
52                  2  
53                  2  
54                  2  
..                ...  
145                 3  
146                 3  
147                 3  
148                 3  
149                 3  

[94 rows x 6 columns]

With **Sort** is possible to sort values in a certain column in an ascending order using `df.sort_values("ColumnName")` or in descending order using `df.sort_values(ColumnName, ascending=False)`. 

Furthermore, it’s possible to sort values by Column1Name in ascending order then Column2Name in descending order by using `df.sort_values([Column1Name,Column2Name],ascending=[True,False])`


In [122]:
df.sort_values("Petal_length") 
#df.sort_values("Height",ascending=False)

Sepal_length  Sepal_width  Petal_length  Petal_width         Species  \
22            4.6          3.6           1.0          0.2     Iris-setosa   
13            4.3          3.0           1.1          0.1     Iris-setosa   
14            5.8          4.0           1.2          0.2     Iris-setosa   
35            5.0          3.2           1.2          0.2     Iris-setosa   
36            5.5          3.5           1.3          0.2     Iris-setosa   
..            ...          ...           ...          ...             ...   
131           7.9          3.8           6.4          2.0  Iris-virginica   
105           7.6          3.0           6.6          2.1  Iris-virginica   
117           7.7          3.8           6.7          2.2  Iris-virginica   
122           7.7          2.8           6.7          2.0  Iris-virginica   
118           7.7          2.6           6.9          2.3  Iris-virginica   

     Species_NumValue  
22                  1  
13                  1  
14                  1  
35                  1  
36                  1  
..                ...  
131                 3  
105                 3  
117                 3  
122                 3  
118                 3  

[150 rows x 6 columns]

The attribute **Groupby** involves splitting the data into groups based on some criteria, applying a function to each group independently and combining the results into a data structure. df.groupby(col) returns a groupby object for values from one column while df.groupby([col1,col2]) returns a groupby object for values from multiple columns.

In [123]:
df.groupby(['Species'])


Size of each group

In [124]:
df.groupby(['Species']).size()

df.groupby(['Species','Species_NumValue']).size()

Species          Species_NumValue
Iris-setosa      1                   50
Iris-versicolor  2                   50
Iris-virginica   3                   50
dtype: int64

This output indicates that we have two types of combinations. 

* Case 1: Species = Iris-setosa & Species_NumValue = 1 
* Case 2: Species = Iris-versicolor & Species_NumValue = 2
* Case 3: Species = Iris-virginica & Species_NumValue = 3

This validates our initial assumption that these three fields essentially portray the same information.

# Data Cleaning: handle with missing data

Before getting started to work with your data, it's a good practice to observe it thoroughly to identify missing values and handle them accordingly.

When reading a dataset using Pandas, there is a set of values including 'NA', 'NULL', and 'NaN' that are taken by default to represent a missing value.  The full list of default missing value codes is in the '`read_csv`' documentation [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html).  This document also explains how to change the way that '`read_csv`' decides whether a variable's value is missing.

Pandas has functions called `isnull` and `notnull` that can be used to identify where the missing and non-missing values are located in a data frame.  

Below we use these functions to count the number of missing and non-missing values in each variable of the datasetr.

In [125]:
df.isnull()
#No null values, indicated by False value

Sepal_length  Sepal_width  Petal_length  Petal_width  Species  \
0           False        False         False        False    False   
1           False        False         False        False    False   
2           False        False         False        False    False   
3           False        False         False        False    False   
4           False        False         False        False    False   
..            ...          ...           ...          ...      ...   
145         False        False         False        False    False   
146         False        False         False        False    False   
147         False        False         False        False    False   
148         False        False         False        False    False   
149         False        False         False        False    False   

     Species_NumValue  
0               False  
1               False  
2               False  
3               False  
4               False  
..                ...  
145             False  
146             False  
147             False  
148             False  
149             False  

[150 rows x 6 columns]

In [126]:
df.notnull()
#No null values, indicated by True value

Sepal_length  Sepal_width  Petal_length  Petal_width  Species  \
0            True         True          True         True     True   
1            True         True          True         True     True   
2            True         True          True         True     True   
3            True         True          True         True     True   
4            True         True          True         True     True   
..            ...          ...           ...          ...      ...   
145          True         True          True         True     True   
146          True         True          True         True     True   
147          True         True          True         True     True   
148          True         True          True         True     True   
149          True         True          True         True     True   

     Species_NumValue  
0                True  
1                True  
2                True  
3                True  
4                True  
..                ...  
145              True  
146              True  
147              True  
148              True  
149              True  

[150 rows x 6 columns]

In [127]:
df.isnull().sum()
#df.notnull().sum()

Sepal_length        0
Sepal_width         0
Petal_length        0
Petal_width         0
Species             0
Species_NumValue    0
dtype: int64

Now we use these functions to count the number of missing and non-missing values in a single variable in the dataset

In [128]:
# Extract all non-missing values of one of the columns into a new variable
x = df.Petal_width.dropna().describe()
x.describe()

count      8.000000
mean      19.745196
std       52.636648
min        0.100000
25%        0.646678
50%        1.249667
75%        1.975000
max      150.000000
Name: Petal_width, dtype: float64

# Add and eliminate columns

In some cases it is useful to create or eiminate new columns 

In [129]:
df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

In [130]:
# Add a new column with new data

# Create a column data
NewColumnData = df.Sepal_length/df.Sepal_length

# Insert that column in the data frame
df.insert(6, "ColumnInserted", NewColumnData, True)

df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  ColumnInserted  
0                 1             1.0  
1                 1             1.0  
2                 1             1.0  
3                 1             1.0  
4                 1             1.0

In [131]:
# Eliminate inserted column
df.drop("ColumnInserted", axis=1, inplace = True)
#df.drop(columns=['ColumnInserted'], inplace = True)
# Remove three columns as index base
#df.drop(df.columns[[12]], axis = 1, inplace = True)

df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

In [132]:
# Add new column derived from existing columns

# The new column is a function of another column
df["PetalLengthToInches"] = df["Petal_length"] * 0.3937

df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  PetalLengthToInches  
0                 1              0.55118  
1                 1              0.55118  
2                 1              0.51181  
3                 1              0.59055  
4                 1              0.55118

In [133]:
# Eliminate inserted column
df.drop("PetalLengthToInches", axis=1, inplace = True)

df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

In [134]:
# Add a new column with text labels reflecting the code's meaning

df["SpeciesGroupNew"] = df.Species_NumValue.replace({1: "Setosa", 2: "Versicolor", 3: "Virginica"})

# Show the first 5 rows of the created data frame
df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue SpeciesGroupNew  
0                 1          Setosa  
1                 1          Setosa  
2                 1          Setosa  
3                 1          Setosa  
4                 1          Setosa

In [135]:
# Eliminate inserted column
df.drop("SpeciesGroupNew", axis=1, inplace = True)
#df.drop(['GenderGroupNew'],vaxis='columns',vinplace=True)

df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

In [136]:
# Add a new column with strata based on these cut points

# Create a column data
NewColumnData = df.Sepal_length/df.Sepal_length

# Insert that column in the data frame
df.insert(1, "ColumnStrata", NewColumnData, True)

df["ColumnStrata"] = pd.cut(df.Petal_width, [60., 63., 66., 69., 72., 75., 78.])

# Show the first 5 rows of the created data frame
df.head()

Sepal_length ColumnStrata  Sepal_width  Petal_length  Petal_width  \
0           5.1          NaN          3.5           1.4          0.2   
1           4.9          NaN          3.0           1.4          0.2   
2           4.7          NaN          3.2           1.3          0.2   
3           4.6          NaN          3.1           1.5          0.2   
4           5.0          NaN          3.6           1.4          0.2   

       Species  Species_NumValue  
0  Iris-setosa                 1  
1  Iris-setosa                 1  
2  Iris-setosa                 1  
3  Iris-setosa                 1  
4  Iris-setosa                 1

In [137]:
# Eliminate inserted column
df.drop("ColumnStrata", axis=1, inplace = True)

df.head()

Sepal_length  Sepal_width  Petal_length  Petal_width      Species  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Species_NumValue  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1

In [82]:
# Drop several "unused" columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
#df.drop(vars, axis=1, inplace = True)

# Add and eliminate rows

In some cases it is requiered to add new observations (rows) to the data set

In [138]:
df.tail()

Sepal_length  Sepal_width  Petal_length  Petal_width         Species  \
145           6.7          3.0           5.2          2.3  Iris-virginica   
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   

     Species_NumValue  
145                 3  
146                 3  
147                 3  
148                 3  
149                 3

In [139]:
df.loc[len(df.index)] = [6.4, 3.7, 4.8, 1.9, 'Iris-virginica', 3] 

df.tail()


Sepal_length  Sepal_width  Petal_length  Petal_width         Species  \
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   
150           6.4          3.7           4.8          1.9  Iris-virginica   

     Species_NumValue  
146                 3  
147                 3  
148                 3  
149                 3  
150                 3

In [100]:
# Eliminate inserted row
df.drop([150], inplace = True )

df.tail()


KeyError: '[150] not found in axis'

# Cleaning your data: drop out unused columns and/or drop out rows with any missing values

In [140]:
# Drop unused columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
#df.drop(vars, axis=1, inplace = True)

#vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
#df = df[vars]

# Drop rows with any missing values
#df = df.dropna()

# Drop unused columns and drop rows with any missing values
#vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
#df = df[vars].dropna()

df


Sepal_length  Sepal_width  Petal_length  Petal_width         Species  \
0             5.1          3.5           1.4          0.2     Iris-setosa   
1             4.9          3.0           1.4          0.2     Iris-setosa   
2             4.7          3.2           1.3          0.2     Iris-setosa   
3             4.6          3.1           1.5          0.2     Iris-setosa   
4             5.0          3.6           1.4          0.2     Iris-setosa   
..            ...          ...           ...          ...             ...   
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   
150           6.4          3.7           4.8          1.9  Iris-virginica   

     Species_NumValue  
0                   1  
1                   1  
2                   1  
3                   1  
4                   1  
..                ...  
146                 3  
147                 3  
148                 3  
149                 3  
150                 3  

[151 rows x 6 columns]

# Final remarks


- The understanding of your dataset is essential
    - Number of observations
    - Variables
    - Data types: numerical or categorial
    - What are my variables of interest

- There are several ways to do the same thing

- Cleaning your dataset (dropping out rows with any missing values) is a good practice

- The **Pandas** library provides fancy, high-performance, easy-to-use data structures and data analysis tools


# Activity: work with the iris dataset

Repeat this tutorial with the iris data set and respond to the following inquiries

1. Calculate the statistical summary for each quantitative variables. Explain the results  

   Better visualization (Below)  

    - Identify the name of each column  
    This data set had no established column names so through research I found what they depicted andd named them: 'Sepal_length', 'Sepal_width', 'Petal_length', 'Petal_width', 'Species', 'Species_NumValue' (a special numeric code for the   species that I created) respectively  
    - Identify the type of each column  
    Sepal_length        float64  
    Sepal_width         float64  
    Petal_length        float64  
    Petal_width         float64  
    Species             object  
    Species_NumValue    int64  
    - Minimum, maximum, mean, average, median, standar deviation  
    Sepal_length	Sepal_width	Petal_length	Petal_width	Species_NumValue  
   count	150.000000	150.000000	150.000000	150.000000	150.000000  
    mean	5.843333	3.057333	3.758000	1.199333	2.000000  
    std     0.828066	0.435866	1.765298	0.762238	0.819232  
    min     4.300000	2.000000	1.000000	0.100000	1.000000  
    max     7.900000	4.400000	6.900000	2.500000	3.000000  
    median  5.8         3.0         4.4         1.3         2.0  

    Explanation: We are analyzing a flower´s (Iris 3 kinds of species) petals and sepals. They are 5.84 x 3.05 cm average on the sepal and 1.19 x 3.75 cm on the petal. We are looking at 150 plants. The smallest sepal is 4.3 cm long and 2 cm wide. The largest sepal is 7.9 cm long and 4.4 cm wide. The smallest petal is 1 cm long and 0.1 wide. The largest petal is 6.9 cm long and 2.5 cm wide.  

2. Are there missing data? If so, create a new dataset containing only the rows with the non-missing data  
    There is no missing data, nor nun nor NaN values in the data set  
3. Create a new dataset containing only the petal width and length and the type of Flower  
    Done below  
    
4. Create a new dataset containing only the setal width and length and the type of Flower  
    Done below  

5. Create a new dataset containing the setal width and length and the type of Flower encoded as a categorical numerical column 
    Done below  


In [142]:
#Excercise 1
df.describe()

Sepal_length  Sepal_width  Petal_length  Petal_width  Species_NumValue
count    151.000000   151.000000    151.000000   151.000000        151.000000
mean       5.847020     3.061589      3.764901     1.203974          2.006623
std        0.826544     0.437548      1.761446     0.761829          0.820542
min        4.300000     2.000000      1.000000     0.100000          1.000000
25%        5.100000     2.800000      1.600000     0.300000          1.000000
50%        5.800000     3.000000      4.400000     1.300000          2.000000
75%        6.400000     3.350000      5.100000     1.800000          3.000000
max        7.900000     4.400000      6.900000     2.500000          3.000000

In [143]:
#Complimentary statistic value for Excercise 1
df.median()

Sepal_length        5.8
Sepal_width         3.0
Petal_length        4.4
Petal_width         1.3
Species_NumValue    2.0
dtype: float64

In [145]:
#Excercise 3
header = ['Petal_length', 'Petal_width']
df.to_csv('Petal_Width_Length.csv', columns = header)

In [146]:
#Excercise 4
header = ['Sepal_length', 'Sepal_width']
df.to_csv('Sepal_Width_Length.csv', columns = header)

In [147]:
#Excercise 5
header = ['Sepal_length', 'Sepal_width', 'Species_NumValue']
df.to_csv('Sepal_Width_Length_Value.csv', columns = header)